File: src/main/kotlin/dev/loveeev/astraTowny/database/MySQL.java
package dev.loveeev.astratowny.database;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import dev.loveeev.astraTowny.Core;
import dev.loveeev.astratowny.config.DatabaseYML;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.logging.Level;

public class MySQL {
    @Getter
    private HikariDataSource dataSource;
    private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();

    public static MySQL getInstance() {
        return instance;
    }

    private static MySQL instance = new MySQL();

    private MySQL() {
        this.initialize();
    }

    public boolean isConnected() {
        return dataSource != null && !dataSource.isClosed();
    }

    private void initialize() {
        HikariConfig config = new HikariConfig();
        config.setJdbcUrl("jdbc:mysql://" + DatabaseYML.getConfig().getString("sql.hostname") + ":" + DatabaseYML.getConfig().getInt("sql.port") + "/" + DatabaseYML.getConfig().getString("sql.dbname"));
        config.setUsername(DatabaseYML.getConfig().getString("sql.username"));
        config.setPassword(DatabaseYML.getConfig().getString("sql.password"));
        config.setPoolName("ASTRATOWNY");

        config.setMaximumPoolSize(DatabaseYML.getConfig().getInt("pooling.max_pool_size"));
        config.setMaxLifetime(DatabaseYML.getConfig().getLong("pooling.max_lifetime"));
        config.setConnectionTimeout(DatabaseYML.getConfig().getLong("pooling.connection_timeout"));

        config.addDataSourceProperty("cachePrepStmts", DatabaseYML.getConfig().getBoolean("pooling.cachePrepStmts"));
        config.addDataSourceProperty("prepStmtCacheSize", DatabaseYML.getConfig().getInt("pooling.prepStmtCacheSize"));
        config.addDataSourceProperty("prepStmtCacheSqlLimit", DatabaseYML.getConfig().getInt("pooling.prepStmtCacheSqlLimit"));
        config.addDataSourceProperty("useServerPrepStmts", DatabaseYML.getConfig().getBoolean("pooling.useServerPrepStmts"));
        config.addDataSourceProperty("autoReconnect", DatabaseYML.getConfig().getBoolean("pooling.autoReconnect"));
        config.addDataSourceProperty("useSSL", DatabaseYML.getConfig().getBoolean("pooling.useSSL"));
        config.addDataSourceProperty("useUnicode", DatabaseYML.getConfig().getBoolean("pooling.useUnicode"));
        config.addDataSourceProperty("characterEncoding", DatabaseYML.getConfig().getString("pooling.characterEncoding"));

        try {
            this.dataSource = new HikariDataSource(config);
        } catch (Exception e) {
            Core.getInstance().getLogger().log(Level.SEVERE, "Ошибка при настройке HikariCP", e);
        }
    }

    public ScheduledExecutorService getExecutor() {
        return this.executorService;
    }

    public Connection getConnection() {
        try {
            if (isConnected()) {
                return this.dataSource.getConnection();
            }
        } catch (SQLException ex) {
            Core.getInstance().getLogger().log(Level.SEVERE, "Ошибка при подключении к базе данных MySQL через HikariCP", ex);
        }
        return null;
    }

}

File: src/main/kotlin/dev/loveeev/astraTowny/database/MySQL.kt
package dev.loveeev.astratowny.database

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import dev.loveeev.astratowny.AstraTowny
import dev.loveeev.astratowny.config.DatabaseYML
import java.sql.Connection
import java.sql.SQLException
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.logging.Level

class MySQL private constructor() {

    private val dataSource: HikariDataSource?
        get() = _dataSource
    private var _dataSource: HikariDataSource? = null

    private val executorService: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()

    companion object {
        private val instance: MySQL = MySQL()

        fun getInstance(): MySQL = instance
    }

    init {
        initialize()
    }

    fun isConnected(): Boolean {
        return dataSource != null && !dataSource!!.isClosed
    }

    private fun initialize() {
        val config = HikariConfig()
        val dbConfig = DatabaseYML.getConfig()
        config.jdbcUrl = "jdbc:mysql://${dbConfig.getString("sql.hostname")}:${dbConfig.getInt("sql.port")}/${dbConfig.getString("sql.dbname")}"
        config.username = dbConfig.getString("sql.username")
        config.password = dbConfig.getString("sql.password")
        config.poolName = "ASTRATOWNY"

        config.maximumPoolSize = dbConfig.getInt("pooling.max_pool_size")
        config.maxLifetime = dbConfig.getLong("pooling.max_lifetime")
        config.connectionTimeout = dbConfig.getLong("pooling.connection_timeout")

        config.addDataSourceProperty("cachePrepStmts", dbConfig.getBoolean("pooling.cachePrepStmts"))
        config.addDataSourceProperty("prepStmtCacheSize", dbConfig.getInt("pooling.prepStmtCacheSize"))
        config.addDataSourceProperty("prepStmtCacheSqlLimit", dbConfig.getInt("pooling.prepStmtCacheSqlLimit"))
        config.addDataSourceProperty("useServerPrepStmts", dbConfig.getBoolean("pooling.useServerPrepStmts"))
        config.addDataSourceProperty("autoReconnect", dbConfig.getBoolean("pooling.autoReconnect"))
        config.addDataSourceProperty("useSSL", dbConfig.getBoolean("pooling.useSSL"))
        config.addDataSourceProperty("useUnicode", dbConfig.getBoolean("pooling.useUnicode"))
        config.addDataSourceProperty("characterEncoding", dbConfig.getString("pooling.characterEncoding"))

        try {
            _dataSource = HikariDataSource(config)
        } catch (e: Exception) {
            AstraTowny.instance.logger.log(Level.SEVERE, "Ошибка при настройке HikariCP", e)
        }
    }

    fun getExecutor(): ScheduledExecutorService {
        return executorService
    }

    fun getConnection(): Connection? {
        return try {
            if (isConnected()) {
                dataSource?.connection
            } else {
                null
            }
        } catch (ex: SQLException) {
            AstraTowny.instance.logger.log(Level.SEVERE, "Ошибка при подключении к базе данных MySQL через HikariCP", ex)
            null
        }
    }
}
